contracts/collect/StepwiseCollectModule.sol

```diff
@@ -0,0 +1,283 @@
+                a: initData.a,
+                referralFee: initData.referralFee,
+                followerOnly: initData.followerOnly,
+                recipient: initData.recipient,
+                b: initData.b,
+                endTimestamp: initData.endTimestamp,
+                c: initData.c,
+                currentCollects: 0,
+                collectLimit: initData.collectLimit
+            });
+            return abi.encode(initData);
+        }
+    }
+
+    /**
+     * @dev Processes a collect by:
+     *  1. Ensuring the collector is a follower
+     *  2. Ensuring the current timestamp is less than or equal to the collect end timestamp
+     *  3. Ensuring the collect does not pass the collect limit
+     *  4. Charging a fee
+     */
+    function processCollect(
+        uint256 referrerProfileId,
+        address collector,
+        uint256 profileId,
+        uint256 pubId,
+        bytes calldata data
+    ) external override onlyHub {
+        if (_dataByPublicationByProfile[profileId][pubId].followerOnly)
+            _checkFollowValidity(profileId, collector);
+        uint256 endTimestamp = _dataByPublicationByProfile[profileId][pubId].endTimestamp;
+        if (endTimestamp != 0 && block.timestamp > endTimestamp) revert Errors.CollectExpired();
+
+        if (
+            _dataByPublicationByProfile[profileId][pubId].collectLimit != 0 &&
+            _dataByPublicationByProfile[profileId][pubId].currentCollects >=
+            _dataByPublicationByProfile[profileId][pubId].collectLimit
+        ) {
+            revert Errors.MintLimitExceeded();
+        } else {
+            ++_dataByPublicationByProfile[profileId][pubId].currentCollects;
+            if (referrerProfileId == profileId) {
+                _processCollect(collector, profileId, pubId, data);
+            } else {
+                _processCollectWithReferral(referrerProfileId, collector, profileId, pubId, data);
+            }
+        }
+    }
+
+    /**
+     * @notice Returns the publication data for a given publication, or an empty struct if that publication was not
+     * initialized with this module.
+     *
+     * @param profileId The token ID of the profile mapped to the publication to query.
+     * @param pubId The publication ID of the publication to query.
+     *
+     * @return ProfilepublicationData The ProfilePublicationData struct mapped to that publication.
+     */
+    function getPublicationData(uint256 profileId, uint256 pubId)
+        external
+        view
+        returns (ProfilePublicationData memory)
+    {
+        return _dataByPublicationByProfile[profileId][pubId];
+    }
+
+    // TODO: Decide if we want keep it "public" or downgrade to "internal"?
+    // Pro-public: Easier frontend calculation/verification of how much you need to pay
+    // Cons-public: extra function call - more gas
+    function calculateFee(ProfilePublicationData memory data) public pure returns (uint256) {
```

Comment:
I'm personally happy with `public` because I think it's a nice utility to be available for frontends.

However it might be weird UX for frontends to form the `data` param given that they'd need to increase `data.currentCollects` by 1 to offset the decrement here. But of course, that decrement does need to be there for the more important function use internally.  So just an interesting UX tradeoff.

Reply:
I think it could be `function calculateFee(uint256 profileId, uint256 pubId) public pure returns (uint256)` so then has the best UX for the frontends. But if we want to use it internally, I guess the only way is having a public one with the sig I mentioned and then an internal one with the `ProfilePublicationData` param. Which I think was the Victor's conclusion in our internal chat

Reply:
Done